locator::topology: add structured datacenter and rack classes

bhalevy opened the pull request:
Instead of index the nodes into racks and datacenters using their names,
keep a map of `datacenter` objects, indexed by name, each holding a map of `rack` objects,
also index by name, each holding an unordered set of `node` objects.

This structured organization of datacenters and racks allows more efficient lookups using naked `const datacenter*` pointers.  Also, this series adds iteration helpers like `topology::for_each_datacenter`, `datacenter::for_each_rack`, and `{datacenter,rack}::for_each_node` to help callers traverse the topology efficiently.

perf-simple-query results:

**Before** (f990ea9678)
median 1478908.07 tps ( 64.3 allocs/op,  17.6 tasks/op,   52804 insns/op,        0 errors)
median absolute deviation: 5123.42
maximum: 1489368.29
minimum: 1408047.93

**After** (44f3d1d196)
median 2157285.39 tps ( 64.1 allocs/op,  17.2 tasks/op,   44715 insns/op,        0 errors)
median absolute deviation: 8909.33
maximum: 2211432.07
minimum: 2127760.95

bhalevy commented:
I have some followup work to not use sstrings in network_topology_strategy _dc_rep_factor which is also used on the hot path.  If we can make sure it is always used along with the token_metadata/topology instance it was created with we could use `datacenter*`, but today we don't guarantee that, so making sure of the latter requires more significant changes.
Alternatively, we can use the datacenter (uu)id, that is valid across shards and topology instances.
I'll give it a shot and see its affect on performance to see if it's worth it.
The scope of that change is much less risky.

bhalevy commented:
Marked as draft since `topology_state_load` is breaking the topology chain (that clones all non-empty or default datacenter id:s) by making a mutable token_metadata from scratch (which creates a new topology instance from scratch).
/cc @kbr-scylla

michoecho commented:
> perf-simple-query results:

What's the exact `perf-simple-query` invocation?

bhalevy commented:
> > perf-simple-query results:
> 
> What's the exact `perf-simple-query` invocation?

default options (none).
after `cpupower frequency-set -g userspace`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6584/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6584/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7826/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6584/)
#### Failed Tests (1/24588):
- [test_network_topology_strategy](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6584/testReport/junit/consistency_test/TestAvailability/Tests___dtest___test_network_topology_strategy) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_network_topology_strategy)
#### Build Details:
- Duration: 2 hr 34 min
- Builder: spider6.cloudius-systems.com

michoecho commented:
> > > perf-simple-query results:
> > 
> > 
> > What's the exact `perf-simple-query` invocation?
> 
> default options (none).

In that case, your results seem to be corrupted, at least the `before` ones. `perf_simple_query` on current master should take about 44k insns/op, not 53k.

bhalevy commented:
> > > > perf-simple-query results:
> > > 
> > > 
> > > What's the exact `perf-simple-query` invocation?
> > 
> > 
> > default options (none).
> 
> In that case, your results seem to be corrupted, at least the `before` ones. `perf_simple_query` on current master should take about 44k insns/op, not 53k.

I'll rebuild and rerun it then.

avikivity commented:
-median 1478908.07 tps ( 64.3 allocs/op, 17.6 tasks/op, 52804 insns/op, 0 errors)
+median 2157285.39 tps ( 64.1 allocs/op, 17.2 tasks/op, 44715 insns/op, 0 errors)

That's just nuts

avikivity commented:
Ah I see it's bogus. Too bad.

avikivity commented on alternator/server.cc:
```diff
@@ -208,7 +208,8 @@ class local_nodelist_handler : public gated_handler {
         // just the list of live nodes in this DC needs more elaborate code:
         auto& topology = _proxy.get_token_metadata_ptr()->get_topology();
         sstring local_dc = topology.get_datacenter();
-        std::unordered_set<gms::inet_address> local_dc_nodes = topology.get_datacenter_endpoints().at(local_dc);
```
Note: the uncaptured get_token_metadata_ptr() above is surely a bug. @nyh 

The type/function for get_token_metadata_ptr() must be changed, it lulls people into a false sense of simplicity.

michoecho replied:
> Note: the uncaptured get_token_metadata_ptr() above is surely a bug.

I'm not seeing it. What do you mean?

bhalevy replied:
it's embedded in `auto& topology = _proxy.get_token_metadata_ptr()->get_topology();`

michoecho replied:
> it's embedded in auto& topology = _proxy.get_token_metadata_ptr()->get_topology();

Yes. What I meant is that the reference isn't used across preemption points, so I didn't see where the bug is. Avi clarified in another comment that it's not a bug now, but a bug waiting to happen.

avikivity commented on alternator/server.cc:
```diff
@@ -208,7 +208,8 @@ class local_nodelist_handler : public gated_handler {
+        const auto& local_dc_nodes = dc_endpoints_map.at(local_dc);
         for (auto& ip : local_dc_nodes) {
             if (_gossiper.is_alive(ip)) {
                 rjson::push_back(results, rjson::from_string(ip.to_sstring()));
```
I guess since we return a ready future below, the lack of capture isn't an active bug, just one waiting to happen.

bhalevy replied:
yes.  we should use a get_token_metadata() method instead to make it clear we're returning a reference to the token_metadata du jour.

avikivity commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -936,7 +936,7 @@ SEASTAR_THREAD_TEST_CASE(test_topology_tracks_local_node) {
     BOOST_REQUIRE_EQUAL(n2->endpoint(), ip2);
-    BOOST_REQUIRE(n2->dc_rack() == endpoint_dc_rack::default_location);
+    BOOST_REQUIRE(n2->dc_rack() == ip1_dc_rack);
 
```
Why do the tests change?

bhalevy replied:
Previously, the nodes kept a `endpoint_dc_rack::default_location` that had to be translated using the topology defaults in its config.
Now, since the nodes keep pointers to datacenter and rack, we instantiate the default datacenter and rack when topology is constructed, and we point to them from the node.location, so we get back the read datacenter and rack names instead of the `default_location` placeholder.

avikivity replied:
So, the tests were enshrining incorrect behavior?

Please add that to the changelog. A patch that changes tests is suspect, so it must prove its innocence.

bhalevy replied:
sure, will add

avikivity commented on locator/topology.hh:
```diff
@@ -166,6 +167,82 @@ private:
+    explicit datacenter(sstring name) noexcept
+        : _name(std::move(name))
+        , _id(datacenter_id::create_random_id())
+    {}
```
Those ids are dangerous, they don't survive restarts and are local to the node.

bhalevy replied:
True, but otherwise we have to use names for data structures that can't guarantee stable or cross shard topology instance (holding the datacenter* and rack* objects)

bhalevy replied:
We could officially keep them in a system.datacenters table as part of consistent topology and then they could be used cluster wide.
In this case the nodes DC and RACK in system.peers could also store those identifiers.

@tgrabiec, @kbr-scylla WDYT?

avikivity replied:
Let's not add new state, it's difficult enough already. This new state would have two sources (snitch + the tables).

avikivity replied:
What we could do is have a global (not thread-local) grow-only std::unordered_set<datacenter_name> and std::unordered_set<rack_name>. We'd then use pointers/references to these objects.

avikivity replied:
I guess this isn't very different from uuids, except that creating new ones is very explicit.

bhalevy replied:
I dropped those identifiers in the latest version.

bhalevy replied:
> What we could do is have a global (not thread-local) grow-only std::unordered_set<datacenter_name> and std::unordered_set<rack_name>. We'd then use pointers/references to these objects.

That's doable [1].  Since topology is built first on shard 0 and cloned to other shards we can just point to the objects on shard 0.

If we leak empty data centers and racks pointers to them can be valid forever.
They can be moved aside or be skipped like this series does. Moving aside would be more efficient if empty DCs/RACKs accumulate over time, but if they are repopulated, we should should probably look them up in the "directory" of empty DCs/RACKs and reuse their objects rather than creating new ones.

[1] A standalone set for `std::unordered_set<rack_name>` won't work since rack names aren't globally unique.
It needs to live in the datacenter scope, or be identified by a `tuple<datacenter_name, rack_name>`.

avikivity commented on locator/topology.hh:
```diff
@@ -166,6 +167,82 @@ private:
+            func(rack);
+            return stop_iteration::no;
+        });
+    }
```
No need to inline this.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -990,26 +987,26 @@ class load_balancer {
-                        if (node.dc_rack().dc == dc) {
-                            rack_load[node.dc_rack().rack] += 1;
+                        if (node.location().dc() == dc) {
+                            rack_load[node.location().rack()] += 1;
```
This asks for short-circuiting location().rack() into rack() and same for dc().

bhalevy replied:
can do

bhalevy replied:
done in latest version

tgrabiec commented on db/consistency_level.cc:
```diff
@@ -115,22 +115,15 @@ bool is_datacenter_local(consistency_level l) {
-    std::unordered_map<sstring, dc_node_count> dc_endpoints;
-    for (auto& dc : nrs->get_datacenters()) {
-        dc_endpoints.emplace(dc, dc_node_count());
-    }
```
This changes the logic, before we would return an entry for each DC even if it didn't have any live endpoints, and now we don't. This may result in assure_sufficient_live_nodes_each_quorum() passing instead of failing.

tgrabiec commented on locator/topology.cc:
```diff
@@ -337,20 +346,38 @@ void topology::index_node(node* node, const endpoint_dc_rack& dr) {
+
+    // Provide strong exception safety guarantees.
+    // No exceptions are allowed after insertion of node succeeded
+    // Note: it is okay to leave behind empty dc/rack since they are ignored
```
Is this path called outside of mutate_token_metadata()? If not, we don't need strong exception guarantees here. If it is, then it shouldn't be because users assume topology held by token_metadata_ptr is immutable.

bhalevy replied:
Even if it is so, I think it's healthy to guarantee those here since it doesn't make the code much more complicated.

tgrabiec replied:
Is it really strongly exception safe though? Before this line we modify members like _nodes_by_endpoint and _datacenters and don't roll back on exception.

bhalevy replied:
You're right, we do rely on the token_metadata/topology immutability so we don't require add_node to be exception safe. I removed that code from the latest version

tgrabiec commented on locator/topology.hh:
```diff
@@ -41,6 +41,35 @@ using node_holder = std::unique_ptr<node>;
+
+public:
+    location() = default;
+    location(const datacenter* dc, const rack* rack) noexcept : _dc(dc), _rack(rack) {}
```
I think it's unsafe to make it a value object and hold unprotected pointers to _dc and _rack. Users may think they're copying location and pass it outside the scope of token_metadata_ptr.

bhalevy replied:
What would you suggest?
To remove the method and to just have node::dc() and node::rack()
Or return datacenter_id and rack_id wrapped in location instead?

tgrabiec replied:
node::dc() and node::rack() are a good idea regardless.

We could also have location() which returns location* and make location non-copyable.

bhalevy replied:
Please see https://github.com/scylladb/scylladb/pull/17273#issuecomment-1943316121 / #17336 for an alternative approach for the dc/rack identifiers.

bhalevy commented:
Please see https://github.com/scylladb/scylladb/pull/17336 as an alternative approach to using either naked pointers or random uuid:s for identification of datacenters and rack.
The gist of it is to use a `frozen_sstring` (new) type that is immutable and pre-calculates a hash of the sstring
which allows efficient comparison for inequality (but adds a size_t compare for equality), and efficient hashing for map / set indexing as key.

bhalevy commented:
In v2 ([00da906](https://github.com/scylladb/scylladb/commit/00da9068469d95a25b0808294b09b566682eee0b)):
- dropped datacenter_id and rack_id
  - consequently dropped last 2 patches off the series
- made class location non-copyable
- added node::dc() and node::rack() methods

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6728/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6728/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7955/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6728/)
#### Build Details:
- Duration: 4 hr 58 min
- Builder: i-0c0afb33658dd5e5b (m5ad.8xlarge)

avikivity commented:
Please update the perf results.

avikivity commented on locator/topology.hh:
```diff
@@ -22,6 +22,7 @@
 
 #include "locator/types.hh"
 #include "inet_address_vectors.hh"
+#include "utils/UUID.hh"
```
unused?

avikivity commented on locator/topology.cc:
```diff
@@ -63,19 +63,51 @@ std::string node::to_string(node::state s) {
+    for (auto dc_it = _datacenters.begin(); dc_it != _datacenters.end(); ) {
+        auto* dc = const_cast<datacenter*>(dc_it->second.get());
+        for (auto rack_it = dc->racks().begin(); rack_it != dc->racks().end(); ) {
+            auto* rack = const_cast<locator::rack*>(rack_it->second.get());
```
Why these const casts? clear_gently isn't const so should have mutable access to everything.

bhalevy replied:
The const_cast:s are needed because the `_datacenters` and `dc->_racks` maps use `const datacenter*` and `const rack*` values, respectively, on purpose, for safety reasons.

bhalevy commented on locator/topology.hh:
```diff
@@ -41,6 +41,35 @@ using node_holder = std::unique_ptr<node>;
+
+    explicit operator bool() const noexcept {
+        return !_dc; // rack is set iff dc is set
+    }
```
dropped from the latest version

bhalevy commented:
In v3 ([feafc29](https://github.com/scylladb/scylladb/commit/feafc293dd944e99108c28be6a9b04154b477a35)):
- most importantly:
  - locator: topology: add topology_registry for datacenters and racks
   - this simplifies the whole series by keeping datacenter and rack objects indefinitely so they can be pointed at without a topology object at hand
 - topology _dc_rack_nodes data type changed to `std::unordered_map<const datacenter*, datacenter_inventory>` where the per-dc inventory consists of an intrusive list of nodes and a map of intrusive list of nodes per-rack.
   - this is more efficient and simple that maintaining `std::unordered_set<const node*>` per dc and per rack.
- Other new patches:
  - locator: topology: node formatter: delete extraneous space char
  - locator: topology: node: set endpoint_dc_rack as part of indexing
  - network_topology_replication_strategy: natural_endpoints_tracker: just count racks and nodes
  - class_registry: move registrator definition into class_registry
  - replication strategies: use central strategy_class_registry definition
  - replication strategies: add per-strategy traits
  - replication strategies: pass topology& to replication strategy ctor
  - consistency_level, storage_proxy: use native locator::datacenter*
  - locator: vnode_effective_replication_map: get_primary_ranges_within_dc: use native datacenter*
  - storage_service: effective_ownership: drop try/catch for get_ranges_for_endpoint
  - storage_service: effective_ownership: fixup indentation
  - locator: vnode_effective_replication_map: get_*_ranges: use native node*
  - service/tablet_allocator: use native datacenter* and rack*
  - service/tablet_allocator: make_plan(dc): use topology inventory
  - locator: topology: return native location/datacenter*/rack*

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6929/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6929/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/90/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8127/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6929/)
#### Failed Tests (4/23366):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6929/testReport/junit/boost.network_topology_strategy_test.test_topology_compare_endpoints.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6929/testReport/junit/boost.network_topology_strategy_test.test_topology_compare_endpoints.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6929/testReport/junit/boost.network_topology_strategy_test.test_topology_compare_endpoints.3.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [test_load_balancing_with_random_load](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6929/testReport/junit/boost/tablets_test/test_load_balancing_with_random_load) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_balancing_with_random_load)
#### Build Details:
- Duration: 2 hr 19 min
- Builder: spider6.cloudius-systems.com

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -925,12 +922,12 @@ class load_balancer {
-                    if (target_node.dc_rack().rack != source_node.dc_rack().rack
-                            && (rack_load[target_node.dc_rack().rack] + 1 > max_rack_load)) {
+                    if (target_node.rack() != source_node.rack()
+                            && (rack_load[target_node.rack()] + 1 > max_rack_load)) {
```
@tgrabiec how did this work previously for nodes in different data centers that happen to use the same rack name?

tgrabiec replied:
target_node and source_node are always in the same datacenter, we don't migrate across DCs.

bhalevy replied:
I see.  Well, I wouldn't count on that here.
The new code is resilient to that.

bhalevy commented:
In v4 ([4f0a712](https://github.com/scylladb/scylladb/commit/4f0a71208d63d2d2e692cc6306106909b3a740db))
- token_metadata: removed stray `#include` and function declaration.
- test/boost/tablets: instantiate topology_registry once (to prevent double registration of metrics), also in storage_proxy_test for symmetry reasons as `topology_registry` is designed to have "static" allocation throughout the program's life time
- network_topology_strategy_test: use `locator.get_dc_rack()` to handle null dc/rack.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6989/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6989/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/129/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8165/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6989/)
#### Build Details:
- Duration: 2 hr 13 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7131/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7131/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8313/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7131/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/275/)
#### Failed Tests (2/23471):
- [test_raft_recovery_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7131/testReport/junit/%28root%29/test_raft_recovery_basic/test_raft_recovery_basic) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_basic)
- [topology_custom.test_raft_recovery_basic.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7131/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_raft_recovery_basic_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_basic.debug.3)
#### Build Details:
- Duration: 2 hr 14 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7131/)
> ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7131/)
> ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8313/)

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8313/testReport/lwt_test/TestLwt/FullDtest___full_split002___test_multi_dc/
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8313/testReport/repair_test/TestRepair/FullDtest___full_split005___test_local_dc_repair/
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8313/testReport/repair_test/TestRepair/FullDtest___full_split005___test_dc_repair/

Those could be related to this change.
Need to investigate further.

> ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7131/)
> ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/275/)
> 
> #### Failed Tests (2/23471):
> * [test_raft_recovery_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7131/testReport/junit/%28root%29/test_raft_recovery_basic/test_raft_recovery_basic) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_basic)
> * [topology_custom.test_raft_recovery_basic.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7131/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_raft_recovery_basic_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_basic.debug.3)

https://github.com/scylladb/scylladb/issues/17662

> 
> #### Build Details:
> * Duration: 2 hr 14 min
> * Builder: spider5.cloudius-systems.com

mykaul commented:
@bhalevy - what's the status of this PR? marked with 'tablets', but it's unclear status.

bhalevy commented:
> @bhalevy - what's the status of this PR? marked with 'tablets', but it's unclear status.

It's severely outdated.  It'll need more than a rebase, possibly rewriting in some parts where the baseline changed significantly.  Since it's an optimization I don't target it for GA.
Given that perf-simple-query doesn't show great advantages, I'm not sure how much of an improvement we would see in real life, so we can decide to shelf this PR indefinitely (sigh, wasted effort, but it is what it is).

bhalevy commented:
Closing as "Won't do" since this approach didn't demonstrate performance gains justifying the change.
